File: FactorialTrailingZeroes172.java
package LeetCodeAlgorithms;

/**
 * Given an integer n, return the number of trailing zeroes in n!.
 * Note: Your solution should be in logarithmic time complexity.
 *
 * Created by dianaluca on 10/6/16.
 */

public class FactorialTrailingZeroes172 {
  public int trailingZeroes(int n) {
    int res = 0;

    while (n != 0) {
      n /= 5;  // complexity: O(log5) logarithm base 5
      res += n; // how many five's contains n.
    }

    return res;
  }
}

File: LeetcodePython/FactorialTrailingZeroes172.py
# Given an integer n, return the number of trailing zeroes in n!.
#
# Note: Your solution should be in logarithmic time complexity.

class Solution(object):
    def trailingZeroesIterative(self, n):
        """
        :type n: int
        :rtype: int
        """
        if n == 0:
            return 0

        cnt = 0
        while n > 0:
            cnt += n//5
            n //= 5

        return cnt

    def trailingZeroesRecursive(self, n):
        """
        :type n: int
        :rtype: int
        """
        if n == 0:
            return 0

        return n / 5 + self.trailingZeroesRecursive(n / 5)